paasta_tools/autoscaling/autoscaling_cluster_lib.py

```diff
@@ -321,26 +340,76 @@ def gracefully_terminate_slave(self, slave_to_kill, current_capacity, new_capaci
     def filter_aws_slaves(self, slaves_list):
         ips = self.get_instance_ips(self.instances, region=self.resource['region'])
         self.log.debug("IPs in AWS resources: {}".format(ips))
-        slaves = [slave for slave in slaves_list if slave_pid_to_ip(slave['task_counts'].slave['pid']) in ips]
+        slaves = [
+            slave for slave in slaves_list
+            if slave_pid_to_ip(slave['task_counts'].slave['pid']) in ips
+        ]
         slave_ips = [slave_pid_to_ip(slave['task_counts'].slave['pid']) for slave in slaves]
-        instance_descriptions = self.describe_instances([], region=self.resource['region'],
-                                                        instance_filters=[{'Name': 'private-ip-address',
-                                                                           'Values': slave_ips}])
+        instance_descriptions = self.describe_instances(
+            instance_ids=[],
+            region=self.resource['region'],
+            instance_filters=[
+                {
+                    'Name': 'private-ip-address',
+                    'Values': slave_ips
+                }
+            ]
+        )
+        instance_statuses = self.describe_instance_status(
+            instance_ids=[],
+            region=self.resource['region'],
+            instance_filters=[
+                {
+                    'Name': 'private-ip-address',
+                    'Values': slave_ips
+                }
+            ]
+        )
         instance_type_weights = self.get_instance_type_weights()
-        slave_instances = [PaastaAwsSlave(slave, instance_descriptions, instance_type_weights) for slave in slaves]
-        return slave_instances
+        paasta_aws_slaves = []
+        for slave in slaves:
+            matching_descriptions = [
+                i for i in instance_descriptions
+                if slave_pid_to_ip(slave['task_counts'].slave['pid'])
+                == i['PrivateIpAddress']
+            ]
+            if matching_descriptions:
+                if len(matching_descriptions) > 1:
+                    log.error("Found more than one instance with the same private IP {0}. "
+                              "This should never happen")
+                    raise ClusterAutoscalingError
+                description = matching_descriptions[0]
+                matching_status = [
+                    status for status in instance_statuses
+                    if status['InstanceId'] == description['InstanceId']
+                ]
+                assert len(status) == 1, "There should be only one InstanceStatus per instance"
```

Comment:
I kinda like the quick and easy assertions for those things that really are bugs and not real life error conditions, feel free to do the same for 377